### PR TITLE
fix: retry ListTopics on transient transport errors

### LINKED ISSUE #3125: ListTopics of clusterAdmin doesn't retry
##### Description

Other methods in ClusterAdmin appear to use the retryOnError mechanism, allowing them to retry upon failure (e.g., CreateTopic, DeleteTopic, CreatePartitions, etc.).
However, ListTopics does not implement any retry logic.

As a result, if the method fails due to reasons like the broker closing the connection (e.g., caused by connections.max.idle.ms), it does not retry. Instead, the client/broker seems to deregister and register again. Once a new broker is registered, the issue no longer occurs.

Is there a specific reason why ListTopics does not include retry logic? Was this behavior intentional?

##### Versions

| Sarama | Kafka | Go  |
| ------ | ----- | --- |
| 1.45.1 |  3.7 | 1.23.0 |

##### Configuration

```go

```

##### Logs

<details><summary>logs: CLICK ME</summary>
<p>

```
time="2025-03-17T17:14:43+09:00" level=debug msg="client/metadata fetching metadata for all topics from broker kafka-broker-test-0:19000\n" 
time="2025-03-17T17:14:43+09:00" level=debug msg="client/metadata got error from broker 0 while fetching metadata: EOF\n" 
time="2025-03-17T17:14:43+09:00" level=debug msg="Closed connection to broker kafka-broker-test-0:19000\n" 
time="2025-03-17T17:14:43+09:00" level=debug msg="client/brokers deregistered broker #0 at kafka-broker-test-0:19000" 
time="2025-03-17T17:14:43+09:00" level=debug msg="client/metadata fetching metadata for all topics from broker kafka-broker-test-1:19001\n" 
time="2025-03-17T17:14:43+09:00" level=debug msg="client/metadata got error from broker 1 while fetching metadata: write tcp 172.24.1.6:53158->10.161.82.44:19001: write: broken pipe\n" 
time="2025-03-17T17:14:43+09:00" level=debug msg="Error while closing connection to broker kafka-broker-test-1:19001: tls: failed to send closeNotify alert (but connection was closed anyway): write tcp 172.24.1.6:53158->10.161.82.44:19001: write: broken pipe\n" 
time="2025-03-17T17:14:43+09:00" level=debug msg="client/brokers deregistered broker #1 at kafka-broker-test-1:19001" 
time="2025-03-17T17:14:43+09:00" level=debug msg="client/metadata fetching metadata for all topics from broker kafka-broker-test-2:19002\n" 
time="2025-03-17T17:14:43+09:00" level=debug msg="client/brokers registered new broker #0 at kafka-broker-test-0:19000" 
time="2025-03-17T17:14:43+09:00" level=debug msg="client/brokers registered new broker #1 at kafka-broker-test-1:19001" 
time="2025-03-17T17:18:05+09:00" level=debug msg="client/metadata fetching metadata for all topics from broker kafka-broker-test-0:19000\n" 
time="2025-03-17T17:18:05+09:00" level=debug msg="Connected to broker at kafka-broker-test-0:19000 (registered as #0)\n" 
time="2025-03-17T17:23:16+09:00" level=debug msg="client/metadata fetching metadata for all topics from broker kafka-broker-test-0:19000\n" 
```

</p>
</details>

### PR DESCRIPTION
**Pull request by huynhanx03**
## Summary

`ClusterAdmin.ListTopics()` currently fails fast on transient connection failures (for example `EOF`, `broken pipe`, timeout) and does not retry like other admin paths.

This change adds retry behavior for `ListTopics` so temporary transport failures can recover automatically.

## Root Cause

`ListTopics` performs two sequential RPCs (`Metadata` then `DescribeConfigs`) but was not wrapped in `retryOnError`, so any transient network failure aborted the whole call immediately.

## Fix

- Wrap the full `ListTopics` flow in `retryOnError` (retry from the start).
- Add a retriable predicate for:
  - existing controller-retriable errors
  - transport-close errors
  - timeout network errors
- Add a regression test that forces one transient metadata request failure, then verifies `ListTopics` retries and succeeds.

## Why This Is Safe

- No behavior change on healthy paths.
- Scope is limited to `ListTopics`.
- Retrying the full flow keeps metadata/config data consistent per attempt.

## Testing

- `go test -run 'TestClusterAdminListTopics|TestClusterAdminListTopicsRetriesOnTransientConnectionError' -count=10 ./...`
- `go test -run TestClusterAdminListTopicsRetriesOnTransientConnectionError -count=20 ./...`
- `go test -run ClusterAdmin -count=1 ./...`

## Issue

Fixes #3125

### CODE REVIEW

File: admin_test.go [hunk 1]
```diff
@@ -193,6 +194,64 @@ func TestClusterAdminListTopics(t *testing.T) {
...
+
+	if len(entries) == 0 {
+		t.Fatal(errors.New("no resource present"))
+	}
+
+	topic, found := entries["my_topic"]
+	if !found {
+		t.Fatal(errors.New("topic not found in response"))
+	}
+	if topic.ReplicaAssignment == nil || topic.ReplicaAssignment[0][0] != 1 {
```
Comment (puellanivis) [inline admin_test.go]: `topic.ReplicaAssignment` is a `map`, so there’s no need to check if it’s `nil` before accessing any keys. A nil map is equivalent to an empty map for reads, which means, every key returns the zero value.

Unlike the map access though, the value of `ReplicaAssignment` is `[]int32`, which means `[0]` _can_ panic, if the slice has length 0.

```suggestion
	assignment := topic.ReplicaAssignment[0]
	if len(assignment) == 0 {
		t.Fatal("ReplicaAssignment[0] was empty")
	}
	if assignment[0] != 1 {
		t.Error("replica assignment was not found in response")
	}
```

**Review by puellanivis — COMMENTED** [review]
Generally looks pretty fine.

**Comment by huynhanx03** [comment]
Hi @puellanivis , thanks for catching that. You're right: the nil map check wasn't useful here, and the real risk was indexing into an empty assignment slice. I've updated the test to check for the partition entry first, then verify the slice is non-empty before asserting the replica ID.

File: admin_test.go [hunk 2]
```diff
@@ -188,7 +189,79 @@ func TestClusterAdminListTopics(t *testing.T) {
...
-	if topic.ReplicaAssignment == nil || topic.ReplicaAssignment[0][0] != 1 {
+	assignment, found := topic.ReplicaAssignment[0]
+	if !found {
+		t.Fatal(errors.New("replica assignment for partition 0 not found in response"))
+	}
+	if len(assignment) == 0 {
+		t.Fatal(errors.New("replica assignment for partition 0 was empty"))
+	}
+	if assignment[0] != 1 {
+		t.Fatal(errors.New("replica assignment not found in response"))
```
Comment (puellanivis) [inline admin_test.go]: The use of `errors.New` here is unnecessary. `t.Fatal` takes `...any`, and a simple string is appropriate, and formats identically.

File: admin_test.go [hunk 3]
```diff
@@ -188,7 +189,79 @@ func TestClusterAdminListTopics(t *testing.T) {
...
+	assignment, found := topic.ReplicaAssignment[0]
+	if !found {
+		t.Fatal(errors.New("replica assignment for partition 0 not found in response"))
+	}
+	if len(assignment) == 0 {
+		t.Fatal(errors.New("replica assignment for partition 0 was empty"))
+	}
+	if assignment[0] != 1 {
 		t.Fatal(errors.New("replica assignment not found in response"))
 	}
```
Comment (puellanivis) [inline admin_test.go]: Same as above, none of these need to be `errors.New`.

**Comment by huynhanx03** [comment]
Hi @puellanivis , I’ve changed those to plain t.Fatal("...").

**Comment by dnwe** [comment]
Hmm, with this PR the new `isRetriableListTopicsError` treats `net.Error.Timeout()` as retryable [here](https://github.com/huynhanx03/sarama/blob/1c31c2416c7e757585524e140ada962fce4dec82/admin.go#L229-L235). However, the retry body reuses the same broker and only calls `b.Open(...)`, ignoring `ErrAlreadyConnected` [here](https://github.com/huynhanx03/sarama/blob/1c31c2416c7e757585524e140ada962fce4dec82/admin.go#L436-L488)

I think this means timeout retries would keep using the same potentially broken connection as `sendAndReceive` only closes the broker when `shouldCloseBrokerConn(err)` [returns true](https://github.com/huynhanx03/sarama/blob/1c31c2416c7e757585524e140ada962fce4dec82/broker.go#L1132-L1149
), but `shouldCloseBrokerConn` explicitly does not close for [timeout]( https://github.com/huynhanx03/sarama/blob/1c31c2416c7e757585524e140ada962fce4dec82/broker.go#L1293-L1322) `net.Error`
 
Once `responseReceiver` hits a read error, it records that as `dead` and [returns the same error](https://github.com/huynhanx03/sarama/blob/1c31c2416c7e757585524e140ada962fce4dec82/broker.go#L1229-L1249) for later promises on that connection. Since `Open` returns `ErrAlreadyConnected` while `opened` is still true, the retry does not [establish a fresh connection](https://github.com/huynhanx03/sarama/blob/1c31c2416c7e757585524e140ada962fce4dec82/broker.go#L160-L168)

So it seems like the new timeout retry path wouldn't recover from the transient timeout case it is trying to handle?
  
I think we need to explicitly close/reopen the broker before retrying a timeout from `GetMetadata` / `DescribeConfigs`?

**Comment by huynhanx03** [comment]
Hi @dnwe , I updated the timeout retry path so `ListTopics` now closes the broker before returning a timeout from both `GetMetadata` and `DescribeConfigs`, allowing the next retry to reopen a fresh connection instead of reusing the same dead one.

I also converted the previous retry test so the timeout is injected after admin initialization, meaning it now exercises the actual `ListTopics` path instead of the client bootstrap metadata refresh. Added coverage for the `DescribeConfigs` timeout case as well.

Tested with: `go test -run 'TestClusterAdminListTopics$|TestClusterAdminListTopicsRetriesOnTransientConnectionError|TestClusterAdminListTopicsRetriesOnDescribeConfigsTimeout' -count=1`

File: admin_test.go [hunk 4]
```diff
@@ -167,29 +168,165 @@ func TestClusterAdminListTopics(t *testing.T) {
...
+	defer seedBroker.Close()
+
+	metadataResponse := NewMockMetadataResponse(t).
+		SetController(seedBroker.BrokerID()).
+		SetBroker(seedBroker.Addr(), seedBroker.BrokerID()).
+		SetLeader("my_topic", 0, seedBroker.BrokerID())
+
+	var metadataAttempts atomic.Int32
+	var injectMetadataTimeout atomic.Bool
+	var metadataTimeoutInjected atomic.Bool
```
Comment (puellanivis) [inline admin_test.go]: Hm… I think the use of atomics here has gotten out of hand. While atomics are good at synchronizing single accesses, two atomic loads and stores do not happen all together atomically.

As a result, there’s a ton of race conditions between tracking number of attempts, and if an error should be injected or not.

The proper non-racey functionality of this code would be better done under a mutex, so that incrementing attempts, and testing for inject timeout can all happen entirely exclusive of setting the trigger to inject timeouts, and getting a baseline of how many times metadata attempts has been called.

**Comment by huynhanx03** [comment]
Hi @puellanivis , I refactored this test to use a mutex-protected state instead of multiple atomics. The timeout injection state, attempt counters, and baseline reads are now synchronized under the same lock, so the test no longer relies on separate atomic operations happening in a particular order. I also cleaned up the retry coverage into a table-driven test for both `MetadataRequest` and `DescribeConfigsRequest` timeout paths.

`go test -run 'TestClusterAdminListTopics$|TestClusterAdminListTopicsRetriesAfterTransientTimeout' -count=1`

File: admin_test.go [hunk 5]
```diff
@@ -167,29 +178,164 @@ func TestClusterAdminListTopics(t *testing.T) {
...
+			injectTimeout := false
+			injected := false
+			metadataAttempts := 0
+			describeConfigsAttempts := 0
+			describeConfigsResponse := NewMockDescribeConfigsResponse(t)
+
+			makeRequestHandler := func(kind adminRequestKind, attempts *int, response func(*request) encoderWithHeader) requestHandlerFunc {
+				return func(req *request) encoderWithHeader {
+					stateMu.Lock()
+					(*attempts)++
```
Comment (puellanivis) [inline admin_test.go]: This function already has access to `metadataAttempts`, there is no need to pass this through a `*int` function parameter.

File: admin_test.go [hunk 6]
```diff
@@ -167,29 +178,164 @@ func TestClusterAdminListTopics(t *testing.T) {
...
+				SetLeader("my_topic", 0, seedBroker.BrokerID())
+
+			var stateMu sync.Mutex
+			injectTimeout := false
+			injected := false
+			metadataAttempts := 0
+			describeConfigsAttempts := 0
+			describeConfigsResponse := NewMockDescribeConfigsResponse(t)
+
+			makeRequestHandler := func(kind adminRequestKind, attempts *int, response func(*request) encoderWithHeader) requestHandlerFunc {
```
Comment (puellanivis) [inline admin_test.go]: `response` is a noun, and thus not well suited for naming a function argument. `respond` would be more appropriate.

File: admin_test.go [hunk 7]
```diff
@@ -5,10 +5,21 @@ package sarama
 import (
 	"errors"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 )
 
+type adminRequestKind string
```
Comment (puellanivis) [inline admin_test.go]: This type is unnecessary. It does not implement any receiver methods, and although there are constants established with this type, I suppose in an attempt to provide a sort of closed enum, that does not provide any type safety in Golang.

An untyped constant string can be assigned into any typed constant string without type conversion, and without syntax error. (Lines 16, and 17 both demonstrates this.)

I can use below `injectOnRequest: "foo",` and nothing will not compile.

File: admin_test.go [hunk 8]
```diff
@@ -5,10 +5,21 @@ package sarama
...
+
+const (
+	metadataRequestKind        adminRequestKind = "MetadataRequest"
+	describeConfigsRequestKind adminRequestKind = "DescribeConfigsRequest"
+	// One injected timeout should force exactly one extra call to complete:
+	// failed call + successful retry.
+	minAttemptsIncreaseAfterInjectedTimeout = 2
+)
```
Comment (puellanivis) [inline admin_test.go]: These constants should be defined within the function body that they are relevant to. (Semantic locality should be reflected in spatial locality.)

File: admin_test.go [hunk 9]
```diff
@@ -5,10 +5,21 @@ package sarama
...
 )
 
+type adminRequestKind string
+
+const (
+	metadataRequestKind        adminRequestKind = "MetadataRequest"
+	describeConfigsRequestKind adminRequestKind = "DescribeConfigsRequest"
+	// One injected timeout should force exactly one extra call to complete:
+	// failed call + successful retry.
+	minAttemptsIncreaseAfterInjectedTimeout = 2
```
Comment (puellanivis) [inline admin_test.go]: One should avoid defining two concepts of consts in one const body. It’s easy enough to specify a new `const` block, or just `const minAttemptsIncreaseAfterInjectedTimeout = 2̀  without a new block.

The idea here is that each definition block should have one conceptual purpose. Listing the request kinds is a group of related constants, and thus reasonable to group. But they’re a separate purpose to `minAttempts…` and so they should be separated syntactically. (Syntax should reaffirm semantics.)

File: admin_test.go [hunk 10]
```diff
@@ -167,29 +178,164 @@ func TestClusterAdminListTopics(t *testing.T) {
...
+	}{
+		{
+			name:            "metadata request",
+			injectOnRequest: metadataRequestKind,
+		},
+		{
+			name:            "describe configs request",
+			injectOnRequest: describeConfigsRequestKind,
+		},
+	}
```
Comment (puellanivis) [inline admin_test.go]: The name described here is merely a repetition of the `injectOnRequest`. Just get ride of it, and use the request kind string:
```suggestion
	tests := []string{
		"MetadataRequest",
		"DescribeConfigsRequest",
	}
```

File: admin_test.go [hunk 11]
```diff
@@ -5,10 +5,21 @@ package sarama
...
+	"sync"
 	"testing"
 	"time"
 )
 
+type adminRequestKind string
+
+const (
+	metadataRequestKind        adminRequestKind = "MetadataRequest"
+	describeConfigsRequestKind adminRequestKind = "DescribeConfigsRequest"
```
Comment (puellanivis) [inline admin_test.go]: These consts serve little purpose. Their value is literally a part of the identifier. This means the identifier is just a longer way to type the same string.

File: admin_test.go [hunk 12]
```diff
@@ -167,29 +178,164 @@ func TestClusterAdminListTopics(t *testing.T) {
...
+			metadataResponse := NewMockMetadataResponse(t).
+				SetController(seedBroker.BrokerID()).
+				SetBroker(seedBroker.Addr(), seedBroker.BrokerID()).
+				SetLeader("my_topic", 0, seedBroker.BrokerID())
+
+			var stateMu sync.Mutex
+			injectTimeout := false
+			injected := false
+			metadataAttempts := 0
+			describeConfigsAttempts := 0
```
Comment (puellanivis) [inline admin_test.go]: There is no need to assign zero values, this is the default initialization value:
```go
// Defining values in this way make explicit the important information: the type, not the irrelevant information: the zero value.
var injectTimeout, injected bool
var metadataAttempts, describeConfigsAttempts int
```

File: admin_test.go [hunk 13]
```diff
@@ -167,29 +178,164 @@ func TestClusterAdminListTopics(t *testing.T) {
...
+					metadataRequestKind,
+					&metadataAttempts,
+					func(req *request) encoderWithHeader { return metadataResponse.For(req.body) },
+				),
+				string(describeConfigsRequestKind): makeRequestHandler(
+					describeConfigsRequestKind,
+					&describeConfigsAttempts,
+					func(req *request) encoderWithHeader { return describeConfigsResponse.For(req.body) },
+				),
+			})
```
Comment (puellanivis) [inline admin_test.go]: I’m unsure why we’re needing to setup these two operations?

We’ve complicated everything by making both tests setup all the test harnesses for the other test at the same time.

In attempting to remove repeating ourselves, we’re instead complicating the tests, and making it less clear what the tests are specifically testing.

We are allowed to go ahead and repeat ourselves a little bit… as a treat.

**Comment by huynhanx03** [comment]
Hi @puellanivis , I simplified the retry tests based on the feedback.

I removed the extra request-kind type, constants, table-driven wrapper, and shared handler helper. The retry coverage is now split into two explicit tests: one for `MetadataRequest` timeout and one for `DescribeConfigsRequest` timeout.

I also switched the test state to idiomatic zero-value declarations under a mutex, so the timeout injection and attempt counters remain synchronized without the extra abstraction.

**Review by puellanivis — COMMENTED** [review]
Looks fine on the code side, I’m unsure if the actual logic fits in with the rest of the functionality. (That is, in the concerns about reusing brokers and under what conditions.)

**Review by dnwe — APPROVED** [review]
LGTM and good enough to merge now. Thanks for working on this and updating it based on the feedback 👍🏻
